pkg/loader/loader.go

```diff
@@ -341,13 +343,87 @@ func LoadRootsWithConfig(cfg *packages.Config, roots ...string) ([]*Package, err
 	// put our build flags first so that callers can override them
 	l.cfg.BuildFlags = append([]string{"-tags", "ignore_autogenerated"}, l.cfg.BuildFlags...)
 
-	rawPkgs, err := packages.Load(l.cfg, roots...)
-	if err != nil {
+	// check each root to see if it should be expanded to include nested modules
+	var goModDirs []string
+	findGoModules := func(p string, d os.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !d.IsDir() && path.Base(p) == "go.mod" {
+			absPathDirOfGoMod, err := filepath.Abs(filepath.Dir(p))
+			if err != nil {
+				return err
+			}
+			goModDirs = append(goModDirs, absPathDirOfGoMod)
+		}
+		return nil
+	}
+	for i, r := range roots {
+		// skip roots whose last path element is not four dot characters
+		if filepath.Base(r) != "...." {
+			continue
+		}
+		// update the root to no longer descend into nested modules
+		roots[i] = r[:len(r)-1]
+		// add any nested modules to the list of Go module directories to
+		// process later
+		if err := filepath.WalkDir(r[:len(r)-4], findGoModules); err != nil {
+			return nil, err
+		}
+	}
+
+	// loadRoots parses the provided file paths and returns their load packages
+	//
+	// uniqueRoots is used to keep track of the discovered packages to be nice
+	// and try and prevent packages from showing up twice when nested module
+	// support is enabled. there is not harm that comes from this per se, but
+	// it makes testing easier when a known number of modules can be asserted
+	uniqueRoots := map[string]struct{}{}
+	loadRoots := func(roots ...string) error {
+		rawPkgs, err := packages.Load(l.cfg, roots...)
+		if err != nil {
+			return err
+		}
+		for _, rawPkg := range rawPkgs {
+			pkg := l.packageFor(rawPkg)
+			if _, found := uniqueRoots[pkg.ID]; !found {
+				l.Roots = append(l.Roots, pkg)
+				uniqueRoots[pkg.ID] = struct{}{}
+			}
+		}
+		return nil
+	}
+
+	// load the packages from the main module
+	if err := loadRoots(roots...); err != nil {
 		return nil, err
 	}
 
-	for _, rawPkg := range rawPkgs {
-		l.Roots = append(l.Roots, l.packageFor(rawPkg))
+	if len(goModDirs) > 0 {
```

Comment:
Nit: Avoid arrowcode through early return, i.E.
```suggestion
	if len(goModDirs) == 0 { return l.roots, nil}
	// submodule handling
```

Reply:
I was not sure what the repository's preference was. I return early in shorter functions but tend to avoid it in longer ones if possible. But I'm happy to follow the project convention.

Reply:
Resolved with latest push.